File: Codes in Python3/Dynamic Programming/Max Product SubArray.py
class Solution:
    def maxProduct(self, nums: List[int]) -> int:
        pre, suff, max_so_far = 0, 0, float('-inf')
        for i in range(len(nums)):
            pre = (pre or 1) * nums[i]
            suff = (suff or 1) * nums[~i]
            max_so_far = max(max_so_far, pre, suff)
        return max_so_far

File: Codes in C++/Dynamic Programming/Max Product SubArray.cpp
class Solution {
public:
    int maxProduct(vector<int>& nums) {
    int pre=1;
    int suff=1;
    int result = INT_MIN;
    for (int i = 0; i < nums.size(); i++) {
        if(pre==0)pre=1;
        if(suff==0)suff=1;
        pre=pre*nums[i];
        suff=suff*nums[nums.size()-i-1];
        result = max(result,max(pre,suff));
    }
    return result;
    }
};